x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/SpanFlyout/DatabaseContext.tsx

```diff
@@ -66,20 +68,25 @@ export function DatabaseContext({ dbContext }: Props) {
       </EuiTitle>
       <EuiSpacer size="m" />
       <DatabaseStatement>
-        <SyntaxHighlighter
-          language={'sql'}
-          style={xcode}
-          customStyle={{
-            color: null,
-            background: null,
-            padding: null,
-            lineHeight: px(unit * 1.5),
-            whiteSpace: 'pre-wrap',
-            overflowX: 'scroll'
-          }}
+        <TruncateHeightSection
+          previewHeight="17.4em"
+          showFull={dbContext.statement.length <= 612}
```

Comment:
These values seem oddly specific?

Reply:
Shouldn't the component itself determine height, and whether to to the full content? 
I think a better api would be `<TruncateHeightSection maxLength={500} />`

Reply:
This would also be nice, but maybe not so easy:
`<TruncateHeightSection maxHeight="300px" />`

Reply:
> Shouldn't the component itself determine height, and whether to show the full content?
> I think a better api would be `<TruncateHeightSection previewLength={500} />`

I agree it would be better, but it would require rendering the statement first, and calculating the height. I was specifically avoiding this by having the consumer (`DatabaseContext`) set the parameters for the appropriate ranges explicitly. 

I'll spend some time and look into doing this automatically, but my intuition tells me it will likely be a complicated solution with `ref` nodes and clientBoundingRect overlap calculation.

Reply:
Would it also require an additional render if you limit it by number of characters?

Reply:
We could also cheat and use something like this https://www.npmjs.com/package/react-animate-height

I think it will be useful for expanding collapsed library frames.

Reply:
My intention with `TruncateHeightSection` was to have it be agnostic to the content it's truncating, so it could be used anywhere. If we want to limit by character count, it must be tightly coupled to `DatabaseContent` since it doesn't render plain text (`SyntaxHighlighter` rendering the formatted db statement with extra nodes for styling). If we're OK with having it be a specific solution for db statements, then character count will work for our purposes.

Reply:
> My intention with `TruncateHeightSection` was to have it be agnostic to the content it's truncating, so it could be used anywhere

Isn't https://www.npmjs.com/package/react-animate-height then a good fit?

Reply:
> We could also cheat and use something like this https://www.npmjs.com/package/react-animate-height
> 
> I think it will also be useful for expanding collapsed library frames. IIRC I used the max-height hack, which has its shortcomings.

`react-animate-height` would be useful if we wanted to animate the expand collapse. There's no animation in this case. (there's no transition on max-height here). Also `react-animate-height` requires pixel height, but what we want is to always show the first 10 lines of the db statement. This solution will always show the first X  number of lines regardless of line height / font-size / etc.

Maybe @formgeist can advise of whether we want animation in the expand/collapse behavior, if we do then we could `react-animate-height` to control the height animation, but we'll still need to calculate the single line height if we want it to always be based on number of lines.

Reply:
Okay, I thought this was all because of the animation. I definitely don't want to make this more complicated, so if we can avoid the animation that's great.

Is the "x number of lines" a hard requirement, or can we limit it by height? If so, I think something as simple as thing can do:
https://codesandbox.io/s/q3v6wq6y39

LMK if I'm oversimplifying this :)

Reply:
(yes, I did use a hook - because we can! Wooop! :D)

Reply:
The requirements don't specify how the height is defined, the design showed the first 10 lines in the preview, we can approximate it by guessing what the height at 10 lines would be and by looking at the character length of the db statement and as to whether or not to show the toggle button, but it will be specific to db statements.

This brings us back to the original concern that the values are oddly specific (https://github.com/elastic/kibana/pull/31674#discussion_r258856563) which my improvements eliminate the need for.

Reply:
Let's clarify with @formgeist if he is okay with a simple solution, like specifying a pixel height or a number of characters. 
Maybe I'm missing something but I think the current implementation is overly complex with the use of refs and recursive whitespace clearing.
What do you think about my codesandbox? Which limitations do you think it has?

Reply:
To be clear, i'm OK with the simpler solution, like in my first pass where i specified the preview height explicity (e.g. `previewHeight="17.4em"`) I received feedback that setting the height like this is too specific and that the component should determine the height itself, which led me to calculate it based on the number of desired preview lines. The codesandbox solution is almost identical to my first pass, which also has a magic number height specific to the view, so it doesn't address the original concern in the first feedback comment. It seems we've now come full circle and landed where we started.

Reply:
Controlling with character count is also a viable alternative for this specific problem, but keep in mind that it's not a generic solution since the text characters are not rendered directly because of the syntax highlighting. I was trying to avoid a specific solution in favor of a generic one in the interest of reuse. But it OK if that does not work in this case.

Reply:
Sorry, I should have been more clear. I think the first solution was fine, apart from the API that required the consumer to specify the `showFull`. I think this should be an implementation detail of the component.

I think it would be great if the component accepted a pixel-height, but that can also be a bit tricky, since we have to calculated whether the content overflows, which is why I suggested that you could cut corners and just count characters. But I agree that it would be a nicer with pixel based solution would be more versatile.
That being said, we are only using this in a single place, and it's hard to know how it will be used, so let's not overthink it.

In this case, I don't think it's strictly required to match the preview height to the line-height. Eg. it's fine to crop the container halfway between a line. In some way this just makes the truncation more obvious to the user.

I updated the codesandbox to make this clear: https://codesandbox.io/s/3y7rm1rzj6

Reply:
FWIW I don't think it's a huge problem if the design needs to define a max character count or max height for an element that will be truncated.